Add SPARC and LEON support

martin-aberg opened the pull request:
# Introduction

This patch set adds support for the SPARC architecture and LEON processors.

`arch/sparc` is compliant with the SPARC V8 specification and independent of processor implementation. This includes SPARC System V ABI compliant software run-time in trap handlers and interrupt context. Number of register windows can be configured.

`soc/sparc` provides support for the LEON3 processors implementation, also compatible with LEON4 and LEON5, including the fault-tolerant versions.

More information about the LEON processors and the SPARC specifications can be found at:
<https://www.gaisler.com/index.php/products/processors>


# Boards

## boards/sparc/qemu_leon3

`qemu_leon3` is a board configuration for exercising the LEON3 port with `sanitycheck`.

This board configuration executes in the QEMU simulator distributed with Zephyr SDK-NG.


## boards/sparc/generic_leon3

`generic_leon3` is a board configuration designed to be used with LEON3 FPGA template designs available in the GRLIB/LEON GPL distribution. The GRLIB GPL distribution can be found at: <https://www.gaisler.com/index.php/downloads/leongrlib>

Sanitycheck can run tests for this board using TSIM3.


## boards/sparc/gr716a_mini

The `gr716a_mini` board configuration targets the GR716A fault-tolerant microcontroller.

Sanitycheck runs tests for this board with the TSIM3 LEON simulator.


# Toolchain

The tools provided by the *Zephyr SDK with Crosstool NG* (SDK-NG) SPARC distribution can be used:

* `sdk-ng v0.11.2`
Works out-of-the-box.

* `sdk-ng v0.11.4` and `v0.12.0-beta-1`
Requires the following SDK-NG patch: <https://github.com/zephyrproject-rtos/sdk-ng/pull/288>. (Replace the file `target.cmake` in the SDK installation.)


# Test results
All tests pass:
```
$ ./scripts/sanitycheck --arch sparc
[...]
INFO    - Total complete:  259/ 259  100%  skipped:   67, failed:    0
INFO    - 192 of 192 tests passed (100.00%), 0 failed, 11768 skipped with 0 warnings in 697.39 seconds
````

tbr-tt committed:
arch: Add SPARC processor architecture

SPARC is an open and royalty free processor architecture.

This commit provides SPARC architecture support to Zephyr. It is
compatible with the SPARC V8 specification and the SPARC ABI and is
independent of processor implementation.

Functionality specific to SPRAC processor implementations should
go in soc/sparc. One example is the LEON3 SOC which is part of this
patch set.

The architecture port is fully SPARC ABI compatible, including trap
handlers and interrupt context.

Number of implemented register windows can be configured.

Some SPARC V8 processors borrow the CASA (compare-and-swap) atomic
instructions from SPARC V9. An option has been defined in the
architecture port to forward the corresponding code-generation option
to the compiler.

Stack size related config options have been defined in sparc/Kconfig
to match the SPARC ABI.

Co-authored-by: Nikolaus Huber <nikolaus.huber.melk@gmail.com>
Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
cmake: Added SPARC support

This commit defines the tool prefix and command line options for the
Zephyr SPARC architectural port.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
intc: Add support for GRLIB IRQMP

This adds support for the GRLIB IRQMP interrupt controller commonly used
in LEON3/4/5 systems.

The driver supports the 15 SPARC interrupts and 16 extended interrupts.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
serial: Add support for GRLIB APBUART

This adds support for the GRLIB APBUART UART peripheral commonly used in
LEON3/4/5 systems.

Driver features:
- Auto-detecting debug FIFO, if configured by GRMON
- Setting and getting UART transfer attributes
- Hardware FIFO if available
- Any number of APBUART devices based on devicetree
- Error status indication
- Polled operation

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
timer: Add support for GRLIB GPTIMER

This adds support for GRLIB GPTIMER general purpose timer used in
LEON3/4/5 systems.

One of the GPTIMER subtimers is used to generate periodic interrutps
for announcing ticks. Another subtimer is used as upcounter for the
cycle_get_32() service.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
soc: LEON3 SPARC V8 Processor

This is a generic SOC description applicable to common LEON3 systems.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
soc: GR716A LEON3FT Microcontroller

This adds support for the GR716A fault-tolerant LEON3 microcontroller.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
sparc: Add compiler and linker tool support

GCC and binutils support for the SPARC architecture.

Co-authored-by: Nikolaus Huber <nikolaus.huber.melk@gmail.com>
Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
sparc: Add testsuite support

This commit provides the timestamp_serialize() define for the SPARC
architecture.

Co-authored-by: Nikolaus Huber <nikolaus.huber.melk@gmail.com>
Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
boards: Add qemu_leon3

This board configuration can be used out-of-the-box with SPARC QEMU
distributed with Zephyr SDK v0.11.2.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
boards: Add GR716-MINI Development Board

GR716A microcontroller evaluation board with features an 80-pin
mezzanine connector. Also supports simulating with TSIM3.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
tests: Big endian support in bitfield test

The BIT_INDEX() macro assumed little-endian. This commit adds
big-endian support, conditioned on the preprocessor define
CONFIG_BIG_ENDIAN.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
tests: kernel: context: SPARC support

This test requires explicit architecture support, which this commit adds
for SPARC.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
tests: Avoid out-of-bounds array access on SPARC

The test reads and writes outside the bounds of an array allocated on
the stack in check_input(). This commit disables the test on SPARC.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
lib/posix: SPARC newlib has unsigned short mode_t

This commit eliminates a compilation error by passing int to va_arg
rather than mode_t on SPARC.

Newlib sys/_types.h defines mode_t for SPARC as:
  typedef unsigned short __mode_t;

GCC 10.2.0 gave the following error message and suggested solution:

mqueue.c: In function 'mq_open':
mqueue.c:61:21: error: 'mode_t' {aka 'short unsigned int'} is promoted
 to 'int' when passed through '...' [-Werror]
   61 |   mode = va_arg(va, mode_t);
      |                     ^
mqueue.c:61:21: note: (so you should pass 'int' not 'mode_t' {aka 'short
unsigned int'} to 'va_arg')

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
libc: minimal: Add SPARC support

This commit defines the off_t type for SPARC.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
sanitycheck: Add TSIM3 support

TSIM3 is a LEON processor and system simulator.

This commit allows using TSIM sanitycheck and the make run target.

TSIM parameters can be further specified in the board configuration
using TSIM and TSIM_SYS.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
boards: Add generic_leon3 for FPGA design

This board configuration is compatible with
- GRLIB GPL FPGA template designs
- TSIM3 LEON3 eval

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
sparc: add support for thread local storage

Adds the necessary bits to initialize TLS in the stack
area and sets up CPU registers during context switch. Register g7 is
used to point to the thread data. Thread data is accessed with negative
offsets from g7.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
logging: change default stack size on SPARC

The default stack size of 768 is just too small for SPARC.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
cmsis: align buffer containing struct k_timer

The slab cmis_timer_slab is used to allocate objects of the type
"struct timer_obj", so it should be defined to use the compile-time
alignment requirement of "struct timer_obj", rather than 4.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
serial: apbuart interrupt support

Adds driver support for APBUART transmitter and receiver
interrupts. Compatible with APBUART implementations without HW FIFO
interrupt, for example QEMU.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
codeowners: add owner for SPARC

Add myself as an owner for SPARC related files.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
cmake: SPARC workaround for SDK toolchain 0.11

Zephyr SDK version 0.11.3, 0.11.4 and 0.12.0-beta-1 does not have
the required SPARC target support.

Zephyr CI is currently using SDK 0.11.3, which causes the following
build error:

CMake Error at /workdir/zephyr/cmake/compiler/gcc/target.cmake:10
(message):
  C compiler /opt/sdk/zephyr-sdk-0.11.3//bin/-gcc not found - Please
  check your toolchain installation

This patch adds a workaround such that the SDK target.cmake is picked
up from the directory 0.11 when ARCH=sparc.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

martin-aberg commented:
The Zephyr CI checks now pass, including simulation of the `qemu_leon3` board configuration.

There were initially an issue related to the Zephyr SDK 0.11.3 CMake files which now has a solution.
* Zephyr SDK `v0.11.3`, `v0.11.4`, `v0.12.0-beta-1`
Depends on Zephyr commit 2ac43a6710c5a013989d30f23267dd37c2b8d9af (part of this pull request) to build for SPARC.

* Zephyr SDK after https://github.com/zephyrproject-rtos/sdk-ng/commit/cff96b0bbafe2774df033a1a82049de33518bb4c
Does not depend on 2ac43a6710c5a013989d30f23267dd37c2b8d9af.

This pull request is compatible with the above Zephyr SDK versions.

andrewboie commented on arch/sparc/Kconfig:
```diff
@@ -0,0 +1,67 @@
+	  Use CASA atomic instructions. Defined by SPARC V9 and available
+	  in some LEON processors.
+
+# The SPARC V8 ABI allocates a stack frame of minimum 96 byte for each SAVE
```
these are some very large stack sizes!

martin-aberg replied:
Yes, stacks are large due to the SPARC ABI stack frames. The stack Kconfig options have been revisited and most of them are reduced in the 2020-11-12 force push.

For example:
```
 config ISR_STACK_SIZE
-       default 8192
+       default 4096

 config TEST_EXTRA_STACKSIZE
-       default 8192 if COVERAGE_GCOV
-       default 4096
+       default 4096 if COVERAGE_GCOV
+       default 2048

 config SYSTEM_WORKQUEUE_STACK_SIZE
-       default 8192
+       default 4096
```

andrewboie commented on arch/sparc/core/interrupt_trap.S:
```diff
@@ -0,0 +1,238 @@
+	beq	.Lno_reschedule
+	 nop
+	/* key (PIL=15) is the first argument */
+	call arch_swap
```
swap is the older, more complex context switch method:

```
/**
 * Cooperatively context switch
 *
 * Must be called with interrupts locked with the provided key.
 * This is the older-style context switching method, which is incompatible
 * with SMP. New arch ports, either SMP or UP, are encouraged to implement
 * arch_switch() instead.
 *
 * @param key Interrupt locking key
 * @return If woken from blocking on some kernel object, the result of that
 *         blocking operation.
 */
int arch_swap(unsigned int key);
```

any particular reason why you didn't implement `arch_switch()`?

martin-aberg replied:
Thank you for the comments regarding `arch_swap()` and `arch_switch()`. I have updated the implementation so it now uses the the `arch_switch()` API instead.

andrewboie replied:
tyvm!

andrewboie commented on include/arch/sparc/arch.h:
```diff
@@ -0,0 +1,119 @@
+}
+
+
+struct __esf {
```
I'm assuming the ESF is so small because the interrupted context's register state gets saved through some other method than pushing it onto a stack? (I don't know this arch well)

martin-aberg replied:
The `struct __esf` type is used only for fatal exceptions in the Zephyr SPARC port. That is, those exceptions which end up in `z_fatal_error()`. So only a minimum of state is saved to the stack because we do not go back from there.

For interrupts on the other hand, more context is saved as required on the stack in `interrupt_trap.S`. The offsets defined in `stack_offsets.h` are used in that case.

andrewboie replied:
If you later add support for CONFIG_USERSPACE, some types of memory protection faults are recoverable. Can cross that bridge when you get there I suppose.

andrewboie commented on samples/subsys/tracing/sample.yaml:
```diff
@@ -23,7 +23,7 @@ tests:
       - CONFIG_OPENOCD_SUPPORT=y
       - CONFIG_SMP=n
-    arch_exclude: posix xtensa
+    arch_exclude: posix xtensa sparc
```
I'd appreciate filing an Enhancement ticket to eventually get tracing working.

martin-aberg replied:
Agreed about filing enhancement ticket to add tracing. Would it be an appropriate way forward to file that ticket after (if) this pull request is accepted?

The `arch_exclude` in `samples/subsys/tracing/sample.yaml` was specific to the `tracing.osawareness.openocd` sample. It was possible to eliminated the `arch_exclude` by providing the corresponding support in `subsys/debug/openocd.c` instead. That update is also covered by the 2020-11-12 force-push.

andrewboie replied:
just open an Enhancement on GitHub summarizing the situation.

martin-aberg replied:
OK!

andrewboie commented on CODEOWNERS:
```diff
@@ -65,6 +65,8 @@
 /soc/x86/                                 @andrewboie @jenmwms @aasthagr
 /arch/xtensa/                             @andrewboie @dcpleung @andyross
 /soc/xtensa/                              @andrewboie @dcpleung @andyross
+/arch/sparc/                              @martin-aberg
```
I would add yourself to the MAINTAINERS file as well.

martin-aberg replied:
Sure. I will have a look at this. What would the appropriate role be,  maintainer or collaborator?

andrewboie replied:
If you're willing to be the maintainer, that would be ideal. We can get you added to the GH contributors list so you can be added to code review. cc @MaureenHelm

MaureenHelm replied:
We need a maintainer, but please start as a collaborator first. You can find more details in the [Project Roles](https://docs.zephyrproject.org/latest/contribute/project_roles.html) document.

martin-aberg replied:
I have now added myself as collaborator in `MAINTAINERS.yml`: b4753af07bc54ad7171be8e740beea210b66f345

tyrantlizard77 reviewed (commented):
This all looks really good. Only nitpick really is the use of old-style `arch_swap()` instead of `arch_switch()`. Can you give some detail on why this was chosen? It may become an issue if SMP is later introduced, and someday we'd like to remove the swap support in the core kernel entirely (not happening anytime soon though).

My review does not cover the ASM code, I don't know this CPU well enough to weigh in. Passing tests are good enough for me, at least.

It looks like we already have a SPARC toolchain present in the SDK, which is great.

tbr-tt committed:
debug: openocd: Support for building on SPARC

This adds the pieces needed for openocd.c to compile when ARCH=SPARC.
In particular, it allows the tracing.osawareness.openocd sample to
build and run.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

nashif commented:
LGTM, can you please some notes in 

doc/releases/release-notes-2.5.rst

So we do not have to chase after you when the release time comes.

tbr-tt committed:
MAINTAINERS: Add entry for SPARC architecture

Add myself as collaborator for the SPARC architecture and its
relevant paths.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
doc: release 2.5: Announce the SPARC support

SPARC architecture is a new feature introduced with Zephyr 2.5.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

martin-aberg commented:
> LGTM, can you please some notes in
> 
> doc/releases/release-notes-2.5.rst

Yes. The notes are added in 1e4eecde8ffe675923585e505aa7ad7615344b5c.

martin-aberg commented:
A couple of string formatting tests have started to fail after #29876 was merged earlier today.

* All tests pass when the SPARC patches are applied on cb1cf54406abafb4a781660161a93d5c363c7a1f.

* The following tests fail when SPARC patches are applied on bb99422c8a5f97d250e3e7c956c57ec78fae0e77:
`tests/lib/sprintf/libraries.libc.sprintf`
`tests/kernel/common/kernel.common.misra`
`tests/kernel/common/kernel.common.tls`
`tests/kernel/common/kernel.common`

I am investigating this right now.

carlescufi commented on arch/Kconfig:
```diff
@@ -37,6 +37,18 @@ config ARM
+	select ARCH_IS_SET
+	select USE_SWITCH
+	select USE_SWITCH_SUPPORTED
+	select BIG_ENDIAN
```
I believe this is our first big endian arch. Are all SPARC ICs big endian, this is not configurable like with Arm?

martin-aberg replied:
Yes, the 32-bit SPARC V8 architecture is always big endian.
(There is also a 64-bit SPARC architecture specification (V9), often referred to as sparc64, which is bi-endian.)

carlescufi commented on boards/sparc/gr716a_mini/doc/index.rst:
```diff
@@ -0,0 +1,216 @@
+In order to build the application for ``gr716a_mini``, set the ``BOARD``
+variable to ``gr716a_mini``.
+
+The applicatoin is linked to the on-chip tightly coupled memory by
```
```suggestion
The application is linked to the on-chip tightly coupled memory by
```

martin-aberg replied:
Thank you. I will correct the typo as suggested.

carlescufi reviewed (approved):
A question and a typo, otherwise this looks great. Thanks!

tbr-tt committed:
lib: cbprintf: improved %p specifier handling

This fixes an issue where the %p specifier always generated "(nil)"
on SPARC. The failing test cases were:
 tests/lib/sprintf/libraries.libc.sprintf
 tests/kernel/common/kernel.common.misra
 tests/kernel/common/kernel.common.tls
 tests/kernel/common/kernel.common

The exact logic behind the issue has not been fully analyzed, but
it can be observed that this commit eliminates one occurrence of
undefined behavior. (Only allowed to read the last union field written.)

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

martin-aberg commented:
> A couple of string formatting tests have started to fail after #29876 was merged earlier today.

This is now fixed in 646d78117104ac1fa5b5ee9b9276224c6bd443c1.

I kept the patch (a one-line removal) in this pull request because the issue was
not reproducible on other platforms before the patch was applied, and the other
platforms still pass all CI tests after the patch is applied.

andrewboie commented:
This doc failure is weird:
```
scripts/gen_kconfig_rest.py: Kconfig.zephyr:23: '/home/runner/work/zephyr/zephyr/doc/_build/Kconfig/Kconfig.shield.defconfig' not found (in 'source "$(KCONFIG_BINARY_DIR)/Kconfig.shield.defconfig"'). Check that environment variables are set correctly (e.g. $srctree, which is set to '/home/runner/work/zephyr/zephyr'). Also note that unset environment variables expand to the empty string.
```
Unrelated to this PR?

martin-aberg commented:
> This doc failure is weird:
> 
> ```
> scripts/gen_kconfig_rest.py: Kconfig.zephyr:23: '/home/runner/work/zephyr/zephyr/doc/_build/Kconfig/Kconfig.shield.defconfig' not found (in 'source "$(KCONFIG_BINARY_DIR)/Kconfig.shield.defconfig"'). Check that environment variables are set correctly (e.g. $srctree, which is set to '/home/runner/work/zephyr/zephyr'). Also note that unset environment variables expand to the empty string.
> ```
> 
> Unrelated to this PR?

Yes it is most likely unrelated to this PR. The only doc change in the latest force push was a typo fix inside a `.rst` file (which was not mentioned in the log message).

There was a time window earlier today in which several PRs failed the _Documentation GitHub Workflow_ CI check in the same way. My understanding is that it was fixed in #30002.

andrewboie commented:
Local doc build with this PR rebased builds fine locally for me.

pabigot commented:
> > A couple of string formatting tests have started to fail after #29876 was merged earlier today.
> 
> This is now fixed in [646d781](https://github.com/zephyrproject-rtos/zephyr/commit/646d78117104ac1fa5b5ee9b9276224c6bd443c1).
> 
> I kept the patch (a one-line removal) in this pull request because the issue was
> not reproducible on other platforms before the patch was applied, and the other
> platforms still pass all CI tests after the patch is applied.

This patch is correct; thanks.  This would have been a residue after late refactoring of pointer handling.
